### PR TITLE
[embedded] Build an initial embedded Swift standard library

### PR DESCRIPTION
**Pull request by kubamracek**
This isn't a "complete" port of the standard library for embedded Swift, but something that should serve as a starting point for further iterations on the stdlib. PR adds a few executable tests that exercise some parts of the stdlib (not any arrays/collections yet, though) in embedded Swift mode -- this achieves actually runnable programs compiled in embedded Swift (!).

- General CMake logic for building a library as ".swiftmodule only" (ONLY_SWIFTMODULE).
- CMake logic in stdlib/public/core/CMakeLists.txt to start building the embedded stdlib for a handful of hardcoded target triples.
- Lots of annotations throughout the standard library to make types, functions, protocols unavailable in embedded Swift (@_unavailableInEmbedded).
- Mainly this is about stdlib functionality that relies on existentials, type erasure, metatypes, reflection, string interpolations.
- We rely on function body removal of unavailable functions to eliminate the actual problematic SIL code (existentials).
- Many .swift files are not included in the compilation of embedded stdlib at all, to simplify the scope of the annotations.
- EmbeddedStubs.swift is used to stub out (as unavailable and fatalError'd) the missing functionality.

### CODE REVIEW

**Comment by kubamracek** [comment]
@swift-ci please test

File: stdlib/public/core/EmbeddedStubs.swift [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+/// Codable
+
+@_unavailableInEmbedded
+public protocol Encodable {
+  func encode(to encoder: any Encoder) throws
+}
+
+@_unavailableInEmbedded
+public protocol Decodable {
+  init(from decoder: any Decoder) throws
```
Comment (phausler) [inline stdlib/public/core/EmbeddedStubs.swift:230]: if we are removing existentials we need to consider `Codable` since that is a requirement for that protocol. Altering the requirements to be `some Decoder` would violate source sharing between embedded and normal swift modes.

This would likely mean that we should consider a potential replacement for both embedded systems and normal systems that avoids existentials (but I don't think you can fully avoid heap allocations via reference types)

File: stdlib/public/core/Equatable.swift [hunk 2]
```diff
@@ -247,10 +247,10 @@ extension Equatable {
 ///   - lhs: A reference to compare.
 ///   - rhs: Another reference to compare.
 @inlinable // trivial-implementation
-public func === (lhs: AnyObject?, rhs: AnyObject?) -> Bool {
+public func ===<T: AnyObject, U: AnyObject>(lhs: T?, rhs: U?) -> Bool {
```
Comment (phausler) [inline stdlib/public/core/Equatable.swift:250]: This isn't a full isomorphism right? because you could have two things stated as `let t: AnyObject = ...` and pass that as `===`.

File: stdlib/public/core/Range.swift [hunk 3]
```diff
@@ -450,9 +462,13 @@ extension Range: Decodable where Bound: Decodable {
           debugDescription: "Cannot initialize \(Range.self) with a lowerBound (\(lowerBound)) greater than upperBound (\(upperBound))"))
     }
     self.init(_uncheckedBounds: (lower: lowerBound, upper: upperBound))
+    #else
+    fatalError()
```
Comment (phausler) [inline stdlib/public/core/Range.swift:466]: Is this just a todo?

File: stdlib/public/core/Equatable.swift (hunk 2 above)
Comment (kubamracek) [reply stdlib/public/core/Equatable.swift:250]: Ah, yes, you're right. Actually some lit tests show this problem, too (on regular Swift), so this is a breaking change. I changed it to #if based alternative implementation only for embedded Swift.

File: stdlib/public/core/Range.swift (hunk 3 above)
Comment (kubamracek) [reply stdlib/public/core/Range.swift]: Fixed and removed, by figuring out how to avoid the #if here altogether: I needed to add an (unavailable) ExpressibleByStringInterpolation conformance to (unavailable) String.

**Comment by kubamracek** [comment]
@swift-ci please test

**Comment by kubamracek** [comment]
@swift-ci please test

**Comment by kubamracek** [comment]
@swift-ci please test

**Comment by kubamracek** [comment]
@swift-ci please test

**Comment by kubamracek** [comment]
@swift-ci please test

**Comment by kubamracek** [comment]
@swift-ci please test

**Comment by kubamracek** [comment]
@swift-ci please test

**Review by phausler — APPROVED** [review]
Overall this looks like a good start to things from the stdlib perspective - there are a few follow-on tasks that we should likely consider in the near future. 

Namely I think we need to address the exposure of `Builtin.NativeObject` with some sort of reference container since the patterns for `Array` are likely used elsewhere and probably deserve a more general solution than relying on builtins.

We also should probably consider some sort of mechanism for steering folks to the right interfaces (via fix-its) when specific requirement configurations are selected. E.g. if it is in the "im targeting something with very low ram requirements" mode (not the non-allocating embedded Swift mode, but something in-between) we should move folks from the say `String.==` to `String.UTF8View.==` due to its potential RAM costs. If this were clang compiling C I would say something named `-Wlarge-allocations`... obviously naming of which should be discussed.

For non-allocating embedded Swift, allocating embedded Swift, and normal Swift we should also consider the utility of quasi replacement types that are removed from the non-allocating embedded variant. Namely `Array` could have some sort of either fixed capacity or fixed size type (without metadata generation for the count).

tl;dr looks good to me

**Review by eeckstein — APPROVED** [review]

**Review by zoecarver — APPROVED** [review]
Awesome! Super excited to see this land!
